src/meta/src/stream/stream_graph.rs

```diff
@@ -184,359 +175,399 @@ impl StreamActorBuilder {
+            }
+
+            // "Leaf" node `Chain`.
+            NodeBody::Chain(chain_node) => {
+                let input = stream_node.get_input();
+                assert_eq!(input.len(), 2);
+
+                let merge_node = &input[0];
+                assert_matches!(merge_node.node_body, Some(NodeBody::Merge(_)));
+                let batch_plan_node = &input[1];
+                assert_matches!(batch_plan_node.node_body, Some(NodeBody::BatchPlan(_)));
+
+                // Index the upstreams by the an external edge ID.
+                let upstreams = &self.upstreams[&EdgeId::UpstreamExternal {
+                    upstream_table_id: chain_node.table_id,
+                    downstream_fragment_id: self.fragment_id,
+                }];
+
+                // As we always use the `NoShuffle` exchange for MV on MV, there should be only one
+                // upstream.
+                let upstream_actor_id = upstreams.actors.as_global_ids();
+                assert_eq!(upstream_actor_id.len(), 1);
+
+                let chain_input = vec![
+                    // Fill the merge node with correct upstream info.
+                    StreamNode {
+                        input: vec![],
+                        stream_key: merge_node.stream_key.clone(),
+                        node_body: Some(NodeBody::Merge(MergeNode {
+                            upstream_actor_id,
+                            upstream_fragment_id: upstreams.fragment_id.as_global_id(),
+                            upstream_dispatcher_type: DispatcherType::NoShuffle as _,
+                            fields: chain_node.upstream_fields.clone(),
+                        })),
+                        fields: chain_node.upstream_fields.clone(),
+                        operator_id: merge_node.operator_id,
+                        identity: "MergeExecutor".to_string(),
+                        append_only: stream_node.append_only,
+                    },
+                    batch_plan_node.clone(),
+                ];
+
+                Ok(StreamNode {
+                    input: chain_input,
+                    identity: "ChainExecutor".to_string(),
+                    fields: chain_node.upstream_fields.clone(),
+                    ..stream_node.clone()
+                })
+            }
+
+            // For other nodes, visit the children recursively.
+            _ => {
+                let mut new_stream_node = stream_node.clone();
+                for (input, new_input) in
+                    stream_node.input.iter().zip_eq(&mut new_stream_node.input)
+                {
+                    *new_input = self.rewrite_inner(input, depth + 1)?;
+                }
+                Ok(new_stream_node)
+            }
         }
     }
+
+    /// Build an actor after all the upstreams and downstreams are processed.
+    pub fn build(self, job: &StreamingJob) -> MetaResult<StreamActor> {
+        let rewritten_nodes = self.rewrite()?;
```

Comment:
Rewrite the node before sealing to `StreamActor`, so that the actor can be immutable once built.